### PR TITLE
#6625 Skip proxy generation for embeddable classes

### LINKED ISSUE #6625: Proxy generation for embedded/transient classes
Due to https://github.com/doctrine/common/pull/771, proxy generation will now fail if a final class is used as an embedded.

We should skip proxy generation for abstract classes, mapped superclassesses, embeddables.

### PR DESCRIPTION
**Pull request by issei-m**
Fix #6625

### CODE REVIEW

**Comment by jaikdean** [comment]
This works for me when using a `final` class as an embeddable 👍  Look forward to it being merged and released!

File: tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php [hunk 1]
```diff
@@ -75,6 +75,26 @@ public function testReferenceProxyDelegatesLoadingToThePersister()
...
     }
 
+    public function testSkipMappedSuperClassesOnGeneration()
+    {
+        $cm = new ClassMetadata(\stdClass::class);
+        $cm->isMappedSuperclass = true;
+
+        $num = $this->proxyFactory->generateProxyClasses([$cm]);
+
+        $this->assertEquals(0, $num, "No proxies generated.");
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php]: Please use `self::assertSame()` instead

File: tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php [hunk 2]
```diff
@@ -75,6 +75,26 @@ public function testReferenceProxyDelegatesLoadingToThePersister()
...
+    }
+
+    public function testSkipEmbeddableClassesOnGeneration()
+    {
+        $cm = new ClassMetadata(\stdClass::class);
+        $cm->isEmbeddedClass = true;
+
+        $num = $this->proxyFactory->generateProxyClasses([$cm]);
+
+        $this->assertEquals(0, $num, "No proxies generated.");
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php]: Please use `self::assertSame()` instead

File: tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php [hunk 3]
```diff
@@ -75,6 +75,26 @@ public function testReferenceProxyDelegatesLoadingToThePersister()
         $proxy->getDescription();
     }
 
+    public function testSkipMappedSuperClassesOnGeneration()
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php]: Please add `@group 6625` on the docblock

File: tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php [hunk 4]
```diff
@@ -75,6 +75,26 @@ public function testReferenceProxyDelegatesLoadingToThePersister()
...
+    {
+        $cm = new ClassMetadata(\stdClass::class);
+        $cm->isMappedSuperclass = true;
+
+        $num = $this->proxyFactory->generateProxyClasses([$cm]);
+
+        $this->assertEquals(0, $num, "No proxies generated.");
+    }
+
+    public function testSkipEmbeddableClassesOnGeneration()
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php]: Please add `@group 6625` on the docblock

File: tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php (hunk 3 above)
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php]: And add ` : void`

File: tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php (hunk 4 above)
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php]: And add ` : void`

**Review by lcobucci — CHANGES_REQUESTED** [review]
LGTM just some small nitpicking 😄

File: tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php (hunk 3 above)
Comment (issei-m) [inline tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php]: > Please add @group 6625 on the docblock

Although this test case is unrelated with #6625 actually? I just found that test case is missing, thus I added. :)

Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php]: True, sorry 😉

**Review by lcobucci — APPROVED** [review]

**Comment by issei-m** [comment]
@lcobucci Fixed!

**Comment by issei-m** [comment]
@lcobucci btw, should this change be shipped to 2.5 branch too, right? The bug seems appeared in 2.5.6.

**Comment by Ocramius** [comment]
@issei-m yes, I'm currently doing that.

**Comment by Ocramius** [comment]
Rebased and manually merged, thanks @issei-m!

`master`: doctrine/doctrine2@9005c5afdde7af70cb64ab647de244b9305da531
`2.5`: doctrine/doctrine2@6184343fd5b4329f7c89910fe49906da2fb74e38
